modules/database/database.bal

```diff
@@ -863,17 +880,59 @@ public isolated transactional function getTimelineSubsteps(int stepId) returns T
     }
 }
 
+public isolated transactional function getTimelineSubstep(int stepId, int substepNr) returns TimelineSubstepSQL|error {
+    stream<TimelineSubstepSQL, sql:Error?> substeps = experimentDB->query(
+        `SELECT stepId, substepNr, substepId, href, hrefUi, cleared FROM TimelineSubstep WHERE stepId=${stepId} AND substepNr=${substepNr};`
+    );
+    var result = substeps.next();
+    check substeps.close();
+
+    // TODO: get associated input data
+    if result is record {|TimelineSubstepSQL value;|} {
+        return result.value;
+    } else if result is error {
+        return result;
+    } else {
+        return error("Could not find timeline substep with step id " + stepId.toString() + " and substep number " + substepNr.toString());
+    }
+}
+
+// TODO
+public isolated transactional function getTimelineSubstepWithParams(int stepId, int substepNr) returns TimelineSubstepWithParams|error {
+    // as in getTimelineStep
+    stream<TimelineSubstepSQL, sql:Error?> substeps = experimentDB->query(
+        `SELECT stepId, substepNr, substepId, href, hrefUi, cleared FROM TimelineSubstep WHERE stepId=${stepId} AND substepNr=${substepNr};`
+    );
+    var result = substeps.next();
+    check substeps.close();
+
+    if result is record {|TimelineSubstepSQL value;|} {
+        return {
+            stepId: result.value.stepId,
+            substepNr: result.value.substepNr,
+            substepId: result.value.substepId,
+            href: result.value.href,
+            hrefUi: result.value.hrefUi,
+            cleared: result.value.cleared,
+            parameters: "" // TODO: how to get those??? -> maybe need to change Step table in plugin runner
```

Comment:
parameters need to also be part found in the substep table as substeps get their own input parameters. (and they are in the sqlite schema)
This is also missing the parameters content type!